relayer/pkg/starknet/chain_client.go

```diff
@@ -0,0 +1,106 @@
+package starknet
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/NethermindEth/juno/core/felt"
+	starknetrpc "github.com/NethermindEth/starknet.go/rpc"
+	gethrpc "github.com/ethereum/go-ethereum/rpc"
+)
+
+// type alias for readibility
+type FinalizedBlock = starknetrpc.Block
+
+// used to create batch requests
+type StarknetBatchBuilder interface {
+	RequestBlockByHash(h *felt.Felt) (StarknetBatchBuilder, error)
+	// RequestBlockByNumber(id uint64) (StarknetBatchBuilder, error)
+	// RequestLatestPendingBlock() (StarknetBatchBuilder, error)
+	// RequestLatestBlockHashAndNumber() (StarknetBatchBuilder, error)
+	// RequestEventsByFilter(f starknetrpc.EventFilter) (StarknetBatchBuilder, error)
+	// RequestTxReceiptByHash(h *felt.Felt) (StarknetBatchBuilder, error)
+	Build() []gethrpc.BatchElem
+}
+
+var _ StarknetBatchBuilder = (*batchBuilder)(nil)
+
+type batchBuilder struct {
+	args []gethrpc.BatchElem
+}
+
+func NewBatchBuilder() StarknetBatchBuilder {
+	return &batchBuilder{
+		args: make([]gethrpc.BatchElem, 0),
+	}
+}
+
+func (b *batchBuilder) RequestBlockByHash(h *felt.Felt) (StarknetBatchBuilder, error) {
+	b.args = append(b.args, gethrpc.BatchElem{
+		Method: "starknet_getBlockWithTxs",
+		Args: []interface{}{
+			starknetrpc.BlockID{Hash: h},
+		},
+		Result: &FinalizedBlock{},
+	})
+	return b, nil
+}
+
+func (b *batchBuilder) Build() []gethrpc.BatchElem {
+	return b.args
+}
+
+type StarknetChainClient interface {
```

Comment:
How different is this from using what we have in `client.go`/`Reader`? We should be able to share the interfaces

Reply:
by share interface do you mean re-use existing methods when possible? there's definitely some possibility for re-using the reader client methods. but if you mean sharing the interfaces we've made it a design decision that chain client interface is distinct and dictated by the customer (atlas now) needs.

Reply:
For this PR specifically, we introduced BlockByHash and BlockHashAndNumber. 

In terms of replacing, re-using the existing calls, I didn't remove/replace BlockByHash (returns txs) with the already written BlockWithTxHashes (returns tx hashes) because we need the block with txs.

I can simplify the logic of LatestBlockHeight to call BlockHashAndNumber, though. Would you like me to do that?